proto/spire/plugin/server/credentialcomposer/v1/credentialcomposer.proto

```diff
@@ -0,0 +1,179 @@
+    // The attributes for the X509 CA. If not included in the response, the
+    // default attributes sent in the request will be used.
+    X509CAAttributes attributes = 1;
+}
+
+message ComposeServerX509SVIDRequest {
+    // The default attributes for the server X509-SVID.
+    X509SVIDAttributes attributes = 1;
+}
+
+message ComposeServerX509SVIDResponse {
+    // The attributes for the X509-SVID. If not included in the response, the
+    // default attributes sent in the request will be used.
+    X509SVIDAttributes attributes = 1;
+}
+
+message ComposeAgentX509SVIDRequest {
+    // The default attributes for the server X509-SVID.
+    X509SVIDAttributes attributes = 1;
+}
+
+message ComposeAgentX509SVIDResponse {
+    // The attributes for the X509-SVID. If not included in the response, the
+    // default attributes sent in the request will be used.
+    X509SVIDAttributes attributes = 1;
+}
+
+message ComposeWorkloadX509SVIDRequest {
+    // The default attributes for the X509-SVID.
+    X509SVIDAttributes attributes = 1;
+
+    // The SPIFFE ID of the workload.
+    string spiffe_id = 2;
+
+    // PKIX encoded public key of the workload.
+    bytes public_key = 3;
+}
+
+message ComposeWorkloadX509SVIDResponse {
+    // The attributes for the X509-SVID. If not included in the response, the
+    // default attributes sent in the request will be used.
+    X509SVIDAttributes attributes = 1;
+}
+
+message ComposeWorkloadJWTSVIDRequest {
+    // The default attributes for the JWT-SVID.
+    JWTSVIDAttributes attributes = 1;
+
+    // The SPIFFE ID of the workload.
+    string spiffe_id = 2;
+}
+
+message ComposeWorkloadJWTSVIDResponse {
+    // The attributes for the JWT-SVID. If not included in the response, the
+    // default attributes sent in the request will be used.
+    JWTSVIDAttributes attributes = 1;
+}
+
+message X509CAAttributes {
+    // The subject of the X509 CA.
+    DistinguishedName subject = 1;
+
+    // Zero or more policy identifiers (OIDs) to apply to the CA.
+    repeated string policy_identifiers = 2;
+
+    // Zero or more extensions to apply to the X509 CA. These will override
+    // any extensions otherwise added by the other fields.
+    repeated X509Extension extra_extensions = 3;
+}
+
+message X509SVIDAttributes {
+    // The subject of the X509-SVID.
+    DistinguishedName subject = 1;
+
+    // Zero or more DNS SANs to apply to the X509-SVID.
+    repeated string dns_sans = 2;
+
+    // Zero or more extensions to apply to the X509-SVID . These will override
+    // any extensions otherwise added by the other fields.
+    repeated X509Extension extra_extensions = 3;
+}
+
+message X509Extension {
+    // The OID of the X.509 extension (e.g. "1.2.3.4")
+    string oid = 1;
+
+    // Opaque value of the extension. No validity checking is performed on
+    // this value. Plugin implementors must ensure they are providing well
+    // formed values for the given extension OID.
```

Comment:
I think it can add some confusion, `No validity checking is performed` are not we going to validate that it is not overriden something that is not expected to update?

Reply:
I just mean that SPIRE server isn't going to try and determine if the value is valid ASN.1, since it really doesn't know what valid values are for arbitrary extensions. I'm not too concerned about overriden extensions, since we'll still validate that this produces a valid X509-SVID...